File: crates/engine/src/tools/web_fetch.rs
use super::web_cache;
use super::web_shared::{domain_pattern, next_user_agent, truncate_output, ParsedHtml};
use super::{str_arg, Tool, ToolContext, ToolFuture, ToolResult};
use base64::Engine;
use serde_json::Value;
use std::collections::HashMap;
use std::time::Duration;

const MAX_RESPONSE_SIZE: usize = 5 * 1024 * 1024;
const DEFAULT_TIMEOUT: u64 = 30;
const MAX_TIMEOUT: u64 = 120;
const MAX_OUTPUT_LINES: usize = 2000;
const MAX_OUTPUT_BYTES: usize = 50 * 1024;

const IMAGE_TYPES: &[&str] = &[
    "image/png",
    "image/jpeg",
    "image/gif",
    "image/webp",
    "image/bmp",
    "image/tiff",
];

pub struct WebFetchTool;

impl Tool for WebFetchTool {
    fn name(&self) -> &str {
        "web_fetch"
    }

    fn description(&self) -> &str {
        "Fetch a URL and extract relevant content using the given prompt. The page is fetched, converted to markdown, then an isolated LLM call extracts only what the prompt asks for."
    }

    fn parameters(&self) -> Value {
        serde_json::json!({
            "type": "object",
            "properties": {
                "url": {
                    "type": "string",
                    "description": "The URL to fetch (must start with http:// or https://)"
                },
                "prompt": {
                    "type": "string",
                    "description": "What to extract or answer from the page content"
                },
                "format": {
                    "type": "string",
                    "enum": ["markdown", "text", "html"],
                    "description": "Output format. Default: markdown"
                },
                "timeout": {
                    "type": "integer",
                    "description": "Timeout in seconds (max 120). Default: 30"
                }
            },
            "required": ["url", "prompt"]
        })
    }

    fn needs_confirm(&self, args: &HashMap<String, Value>) -> Option<String> {
        let url = str_arg(args, "url");
        Some(url.to_string())
    }

    fn approval_patterns(&self, args: &HashMap<String, Value>) -> Vec<String> {
        // Only generate a pattern for the requested URL's domain. If the
        // response redirects to a different domain we re-prompt rather than
        // silently following (see execute / fetch_raw).
        domain_pattern(&str_arg(args, "url")).into_iter().collect()
    }

    fn execute<'a>(
        &'a self,
        args: HashMap<String, Value>,
        ctx: &'a ToolContext<'a>,
    ) -> ToolFuture<'a> {
        Box::pin(async move {
            // Fetch the page (blocking IO)
            let raw = tokio::task::block_in_place(|| self.fetch_raw(&args));
            if raw.is_error {
                return raw;
            }

            // Extract content using LLM
            let prompt = str_arg(&args, "prompt");
            match ctx
                .provider
                .extract_web_content(&raw.content, &prompt, ctx.model)
                .await
            {
                Ok(extracted) => ToolResult::ok(extracted),
                Err(_) => raw,
            }
        })
    }
}

impl WebFetchTool {
    fn fetch_raw(&self, args: &HashMap<String, Value>) -> ToolResult {
        let url_str = str_arg(args, "url");
        let format = str_arg(args, "format");
        let format = if format.is_empty() {
            "markdown".to_string()
        } else {
            format
        };
        let timeout_secs = args
            .get("timeout")
            .and_then(|v| v.as_u64())
            .unwrap_or(DEFAULT_TIMEOUT)
            .min(MAX_TIMEOUT);

        if !url_str.starts_with("http://") && !url_str.starts_with("https://") {
            return ToolResult::err("URL must start with http:// or https://");
        }

        let parsed_url = match url::Url::parse(&url_str) {
            Ok(u) => u,
            Err(e) => return ToolResult::err(format!("Invalid URL: {e}")),
        };

        let cache_key = format!("fetch:{url_str}:{format}");
        if let Some(cached) = web_cache::get(&cache_key) {
            return ToolResult::ok(cached);
        }

        let fetch_url = url_str.clone();
        let fetch_result = std::thread::spawn(move || {
            let timeout = Duration::from_secs(timeout_secs);
            let ua = next_user_agent();
            let client = reqwest::blocking::Client::builder()
                .timeout(timeout)
                .user_agent(ua)
                .redirect(reqwest::redirect::Policy::limited(10))
                .build()?;

            let response = client
                .get(&fetch_url)
                .header(
                    "Accept",
                    "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                )
                .header("Accept-Language", "en-US,en;q=0.9")
                .send()?;

            let response = if response.status().as_u16() == 403 {
                let cf = response
                    .headers()
                    .get("cf-mitigated")
                    .and_then(|v| v.to_str().ok())
                    .unwrap_or("");
                if cf == "challenge" {
                    client
                        .get(&fetch_url)
                        .header("User-Agent", "agent")
                        .header("Accept", "text/html,application/xhtml+xml,*/*;q=0.8")
                        .header("Accept-Language", "en-US,en;q=0.9")
                        .send()?
                } else {
                    response
                }
            } else {
                response
            };

            Ok::<_, reqwest::Error>(response)
        })
        .join();

        let response = match fetch_result {
            Ok(Ok(r)) => r,
            Ok(Err(e)) => return ToolResult::err(format!("Fetch failed: {e}")),
            Err(_) => return ToolResult::err("Fetch thread panicked"),
        };

        // Check if redirects landed on a different domain than requested.
        let final_url = response.url().clone();
        if final_url.host_str() != parsed_url.host_str() {
            return ToolResult::err(format!(
                "Redirect crossed domains: requested {} but landed on {}. \
                 Re-run with the final URL if intended.",
                parsed_url.host_str().unwrap_or("?"),
                final_url.host_str().unwrap_or("?"),
            ));
        }

        let status = response.status();
        if !status.is_success() {
            return ToolResult::err(format!("HTTP {status}"));
        }

        let content_type = response
            .headers()
            .get("content-type")
            .and_then(|v| v.to_str().ok())
            .unwrap_or("")
            .to_lowercase();

        if IMAGE_TYPES.iter().any(|t| content_type.contains(t)) {
            use std::io::Read;
            let mut buf = Vec::new();
            if let Err(e) = response
                .take(MAX_RESPONSE_SIZE as u64)
                .read_to_end(&mut buf)
            {
                return ToolResult::err(format!("Failed to read image: {e}"));
            }
            let b64 = base64::engine::general_purpose::STANDARD.encode(&buf);
            let mime = content_type.split(';').next().unwrap_or("image/png").trim();
            return ToolResult::ok(format!("![image](data:{mime};base64,{b64})"));
        }

        let (body, was_truncated) = {
            use std::io::Read;
            let cap = response
                .content_length()
                .map(|l| (l as usize).min(MAX_RESPONSE_SIZE + 1))
                .unwrap_or(MAX_RESPONSE_SIZE + 1);
            let mut buf = Vec::with_capacity(cap);
            if let Err(e) = response
                .take(MAX_RESPONSE_SIZE as u64 + 1)
                .read_to_end(&mut buf)
            {
                return ToolResult::err(format!("Failed to read response: {e}"));
            }
            let truncated = buf.len() > MAX_RESPONSE_SIZE;
            if truncated {
                buf.truncate(MAX_RESPONSE_SIZE);
            }
            (String::from_utf8_lossy(&buf).into_owned(), truncated)
        };

        let is_html = content_type.contains("text/html") || content_type.contains("xhtml");

        // Parse HTML once and extract title, links, and content in a single pass.
        let (title, links, content) = if is_html {
            let parsed = ParsedHtml::parse(&body, Some(&parsed_url));
            let content = match format.as_str() {
                "text" => parsed.to_text(),
                "html" => body.clone(),
                _ => parsed.to_markdown(),
            };
            (parsed.title, parsed.links, content)
        } else {
            (None, vec![], body.clone())
        };

        let mut output = String::new();
        if let Some(t) = &title {
            output.push_str(&format!("# {t}\n\n"));
        }
        output.push_str(&content);
        if !links.is_empty() {
            output.push_str("\n\n## Links\n\n");
            for link in &links {
                output.push_str(&format!("- {link}\n"));
            }
        }

        let mut output = truncate_output(&output, MAX_OUTPUT_LINES, MAX_OUTPUT_BYTES);
        if was_truncated {
            output.push_str("\n\n[Response truncated — original response exceeded 5 MB]");
        }
        web_cache::put(&cache_key, &output);

        ToolResult::ok(output)
    }
}

File: runtime/lua/smelt/tools/web_fetch.lua
-- Built-in web_fetch tool. Fetches a URL and extracts content via an isolated LLM call.

local MAX_RESPONSE_SIZE = 5 * 1024 * 1024
local DEFAULT_TIMEOUT = 30
local MAX_TIMEOUT = 120
local MAX_OUTPUT_LINES = 2000
local MAX_OUTPUT_BYTES = 50 * 1024

local IMAGE_MIMES = {
  "image/png",
  "image/jpeg",
  "image/gif",
  "image/webp",
  "image/bmp",
  "image/tiff",
}

local function url_host(url)
  local _, host = url:match("^([Hh][Tt][Tt][Pp][Ss]?)://([^/?#]+)")
  if host then host = host:lower() end
  return host
end

local function domain_pattern(url)
  local scheme, host = url:match("^([Hh][Tt][Tt][Pp][Ss]?)://([^/?#]+)")
  if not scheme or not host then return nil end
  return scheme:lower() .. "://" .. host:lower() .. "/*"
end

-- Hard line cap, then char-safe byte cap, then a truncation note.
local function truncate_output(text, max_lines, max_bytes)
  local lines = {}
  for line in (text .. "\n"):gmatch("([^\n]*)\n") do
    lines[#lines + 1] = line
  end
  local truncated = false
  if #lines > max_lines then
    while #lines > max_lines do table.remove(lines) end
    truncated = true
  end
  local result = table.concat(lines, "\n")
  if #result > max_bytes then
    local cut = max_bytes
    while cut > 0 do
      local b = result:byte(cut + 1)
      if not b or b < 0x80 or b >= 0xC0 then break end
      cut = cut - 1
    end
    result = result:sub(1, cut)
    truncated = true
  end
  if truncated then
    result = result .. "\n\n[output truncated]"
  end
  return result
end

local function header(headers, key)
  if not headers then return "" end
  return headers[key] or headers[key:lower()] or headers[key:upper()] or ""
end

local function fetch_raw(args)
  local url = args.url or ""
  local format = args.format
  if not format or format == "" then format = "markdown" end
  local timeout = math.min(tonumber(args.timeout) or DEFAULT_TIMEOUT, MAX_TIMEOUT)

  if not url:match("^[Hh][Tt][Tt][Pp][Ss]?://") then
    return { content = "URL must start with http:// or https://", is_error = true }
  end

  local req_host = url_host(url)
  if not req_host then
    return { content = "Invalid URL: " .. url, is_error = true }
  end

  local cache_key = "fetch:" .. url .. ":" .. format
  local cached = smelt.http.cache.read(cache_key)
  if cached then return cached end

  local function do_fetch(ua)
    return smelt.http.get(url, {
      timeout_secs = timeout,
      max_redirects = 10,
      headers = {
        ["User-Agent"] = ua,
        ["Accept"] = "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        ["Accept-Language"] = "en-US,en;q=0.9",
      },
    })
  end

  local resp, err = do_fetch(smelt.http.random_user_agent())
  if err then
    return { content = "Fetch failed: " .. err, is_error = true }
  end
  if resp.status == 403 and header(resp.headers, "cf-mitigated"):lower() == "challenge" then
    resp, err = do_fetch("smelt")
    if err then
      return { content = "Fetch failed: " .. err, is_error = true }
    end
  end

  local final_host = url_host(resp.final_url or url)
  if final_host ~= req_host then
    return {
      content = string.format(
        "Redirect crossed domains: requested %s but landed on %s. "
          .. "Re-run with the final URL if intended.",
        req_host or "?",
        final_host or "?"
      ),
      is_error = true,
    }
  end

  if resp.status < 200 or resp.status >= 300 then
    return { content = "HTTP " .. resp.status, is_error = true }
  end

  local content_type = header(resp.headers, "content-type"):lower()
  for _, mime in ipairs(IMAGE_MIMES) do
    if content_type:find(mime, 1, true) then
      local primary = content_type:match("^([^;]+)") or mime
      primary = primary:gsub("%s+", "")
      local body = resp.body
      if #body > MAX_RESPONSE_SIZE then body = body:sub(1, MAX_RESPONSE_SIZE) end
      local data_url = smelt.image.data_url_from_bytes(body, primary)
      return "![image](" .. data_url .. ")"
    end
  end

  local body = resp.body
  local was_truncated = false
  if #body > MAX_RESPONSE_SIZE then
    body = body:sub(1, MAX_RESPONSE_SIZE)
    was_truncated = true
  end

  local is_html = content_type:find("text/html", 1, true)
    or content_type:find("xhtml", 1, true)

  local title, links, content
  if is_html then
    if format == "html" then
      content = body
      title = smelt.html.title(body)
      links = smelt.html.links(body, url) or {}
    else
      local md = smelt.html.to_markdown(body, url)
      title = md.title
      links = md.links or {}
      content = (format == "text") and smelt.html.to_text(body) or md.content
    end
  else
    title = nil
    links = {}
    content = body
  end

  local parts = {}
  if title and title ~= "" then
    parts[#parts + 1] = "# " .. title .. "\n\n"
  end
  parts[#parts + 1] = content
  if #links > 0 then
    parts[#parts + 1] = "\n\n## Links\n\n"
    for _, link in ipairs(links) do
      parts[#parts + 1] = "- " .. link .. "\n"
    end
  end

  local output = truncate_output(table.concat(parts), MAX_OUTPUT_LINES, MAX_OUTPUT_BYTES)
  if was_truncated then
    output = output .. "\n\n[Response truncated — original response exceeded 5 MB]"
  end
  smelt.http.cache.write(cache_key, output)
  return output
end

-- Spawn an auxiliary LLM request and park the coroutine until it responds.
-- Returns the assistant reply on success, or `nil` on failure (the caller
-- falls back to the raw fetched content).
local function ask_extract(content, prompt)
  local id = smelt.task.alloc()
  smelt.engine.ask({
    system = "Answer the user's question based solely on the provided "
      .. "web page content. Be concise and direct.",
    question = "<content>\n" .. content .. "</content>\n\n" .. prompt,
    model = smelt.model.preferred("web_fetch"),
    on_response = function(resp, err) smelt.task.resume(id, { content = resp, err = err }) end,
  })
  local result = smelt.task.wait(id)
  if not result or result.err then return nil end
  return result.content
end

smelt.tools.register({
  name = "web_fetch",
  description = "Fetch a URL and extract relevant content using the given prompt. "
    .. "The page is fetched, converted to markdown, then an isolated LLM call "
    .. "extracts only what the prompt asks for.",
  override = true,
  permission_defaults = { normal = "allow", plan = "allow", apply = "allow" },
  parameters = {
    type = "object",
    properties = {
      url = {
        type = "string",
        description = "The URL to fetch (must start with http:// or https://)",
      },
      prompt = {
        type = "string",
        description = "What to extract or answer from the page content",
      },
      format = {
        type = "string",
        enum = { "markdown", "text", "html" },
        description = "Output format. Default: markdown",
      },
      timeout = {
        type = "integer",
        description = "Timeout in seconds (max 120). Default: 30",
      },
    },
    required = { "url", "prompt" },
  },
  summary = function(args) return args.url or "" end,
  approval_patterns = function(args)
    local pat = domain_pattern(args.url or "")
    if pat then return { pat } end
    return {}
  end,
  render = function(args, output, ctx)
    local items = {}
    if args.prompt and args.prompt ~= "" then
      table.insert(items, smelt.layout.text(args.prompt))
    end
    table.insert(items, smelt.layout.text(output.content, {
      hl_group = output.is_error and "ErrorMsg" or nil,
    }))
    return smelt.layout.vbox(items)
  end,
  decide = function(args, mode)
    -- Deny dominates; pattern allow short-circuits; Allow tool + Ask pattern → Ask.
    local tool = smelt.permissions.check_tool(mode, "web_fetch")
    if tool == "deny" then return "deny" end
    local pat = smelt.permissions.check(mode, "web_fetch", args.url or "")
    if pat == "deny" then return "deny" end
    if pat == "allow" then return "allow" end
    if tool == "allow" and pat == "ask" then return "ask" end
    return pat
  end,
  execute = function(args)
    local raw = fetch_raw(args)
    if type(raw) == "table" and raw.is_error then return raw end
    local content = type(raw) == "table" and raw.content or raw
    local prompt = args.prompt or ""
    local extracted = ask_extract(content, prompt)
    if extracted and extracted ~= "" then
      return extracted
    end
    return raw
  end,
})
